File: src/main/java/com/letgo/shared/infrastructure/event/publisher/SpringAsyncDomainEventPublisher.java
package com.letgo.shared.infrastructure.event.publisher;

import com.letgo.shared.application.event.DomainEventPublisher;
import com.letgo.shared.domain.DomainEvent;
import com.letgo.shared.infrastructure.InfrastructureService;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.scheduling.annotation.EnableAsync;

import java.util.List;

@EnableAsync
@InfrastructureService
final public class SpringAsyncDomainEventPublisher implements DomainEventPublisher {
    private final ApplicationEventPublisher publisher;

    public SpringAsyncDomainEventPublisher(ApplicationEventPublisher publisher) {
        this.publisher = publisher;
    }

    @Override
    public void publish(List<DomainEvent> events) {
        events.forEach(publisher::publishEvent);
    }
}

File: src/main/java/com/letgo/shared/infrastructure/event/publisher/SpringAsyncDomainEventPublisher.kt
package com.letgo.shared.infrastructure.event.publisher

import com.letgo.shared.application.event.DomainEventPublisher
import com.letgo.shared.domain.DomainEvent
import com.letgo.shared.infrastructure.InfrastructureService
import org.springframework.context.ApplicationEventPublisher
import org.springframework.scheduling.annotation.EnableAsync

@EnableAsync
@InfrastructureService
class SpringAsyncDomainEventPublisher(
    private val publisher: ApplicationEventPublisher,
) : DomainEventPublisher {
    override fun publish(events: List<DomainEvent>) {
        events.forEach { publisher.publishEvent(it) }
    }
}
